pennylane_lightning/core/src/simulators/lightning_gpu/StateVectorCudaManaged.hpp

```diff
@@ -1100,6 +1171,590 @@ class StateVectorCudaManaged
         return -static_cast<PrecisionT>(0.5);
     }
 
+    /* Controlled-gate generators */
+    /**
+     * @brief Gradient generator function associated with the controlled-RX
+     * gate.
+     *
+     * @param controlled_wires Control wires.
+     * @param controlled_values Control values (false or true).
+     * @param wires Target wires to apply operation.
+     * @param adj Takes adjoint of operation if true. Defaults to false.
+     */
+    inline PrecisionT
+    applyControlledGeneratorRX(const std::vector<std::size_t> &controlled_wires,
+                               const std::vector<bool> &controlled_values,
+                               const std::vector<std::size_t> &wires,
+                               bool adj = false) {
+        const std::size_t ctrl_size = controlled_wires.size();
+        const std::size_t tgt_size = wires.size();
+
+        // Generate permutation
+        std::vector<custatevecIndex_t> permutations(
+            Pennylane::Util::exp2(ctrl_size + tgt_size));
+        std::iota(permutations.begin(), permutations.end(), 0);
+        std::size_t index = 0;
+        for (std::size_t i = 0; i < controlled_values.size(); ++i) {
+            index |= controlled_values[i] << (ctrl_size + tgt_size - 1 - i);
+        }
```

Comment:
We are working to ensure our codebase is updated to C++20 and making use of STL functions wherever possible. We don't need to address this in the current PR due to the workload involved, but if you have time, feel free to make the updates as suggested below. In case of proceeding, please double-check my code and make sure to apply these changes to all the other generators.
```suggestion
        std::size_t i = 0;
        std::size_t index = std::accumulate(controlled_values.begin(), controlled_values.end(), std::size_t{0},
        [&i, ctrl_size, tgt_size] (std::size_t acc, bool value) {
                return acc | (static_cast<std::size_t>(value) << (ctrl_size + tgt_size - 1 - i++ ));
        });
```
Note the addition of a `static_cast` to `size_t` in the variable `value` as bitwise shifts and OR expect integer operands.

Reply:
That's wonderful! I've applied that now.